10208: Add a PyNaCl-based fallback path for Ed25519

cjwatson opened the pull request:
## Scope and purpose

Some commonly-used platforms that Twisted still supports, such as Ubuntu 18.04 LTS, lack a version of OpenSSL new enough to support Ed25519 via cryptography.  Add a fallback path using PyNaCl to support these.

I went back and forward on the best way to handle installing PyNaCl.  I think it's probably reasonable to just add it to the `conch` extra as I've done here, because installing it doesn't too onerous (it uses `libsodium`, but has a built-in copy of it, and PyPI has plenty of wheels), and it results in the convenient property that Ed25519 is always supported rather than being conditional.  However, if this turns out to be a problem on some platforms, then we can drop it from the extra, import it conditionally, and just say that you need to install it manually if you don't have a new enough OpenSSL.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10208
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: cjwatson
Reviewer: <github_username>, <github_username_if_more_reviewers>
Fixes: ticket:10208

Add a PyNaCl-based fallback path for Ed25519

Some commonly-used platforms that Twisted still supports, such as Ubuntu
18.04 LTS, lack a version of OpenSSL new enough to support Ed25519 via
cryptography.  Add a fallback path using PyNaCl to support these.
```

cjwatson committed:
Add a PyNaCl-based fallback path for Ed25519

Some commonly-used platforms that Twisted still supports, such as Ubuntu
18.04 LTS, lack a version of OpenSSL new enough to support Ed25519 via
cryptography.  Add a fallback path using PyNaCl to support these.

cjwatson committed:
Fix up typing, more or less

cjwatson committed:
Simplify twisted.conch.test.test_keys

Since Ed25519 is now always supported, we don't need all these
conditionals.

cjwatson committed:
Add more tests for details of PyNaCl shims

adiroiban commented on setup.cfg:
```diff
@@ -83,6 +83,7 @@ conch =
     cryptography >= 2.6
     appdirs >= 1.4.0
     bcrypt >= 3.0.0
+    PyNaCl
```
Maybe add a separate target.

Most users will have a modern cryptography and nacl is not needd

```suggestion
conch-nacl:
    %(conch)s
    PyNaCl
```

cjwatson replied:
Seems reasonable.  Done (not quite via your suggestion, but similar).

adiroiban commented:
Thanks for the PR and sorry for the late reply. You can ping me for any SSH related PR and I will try to make it a top priority.

I guess that this is related to https://github.com/twisted/twisted/pull/1605

--------

I think that using pynacl is the right solution.

I would prefer to have pynacl as a soft-dependency so that users with a modern openssl+crytography don't have to worry about pynacl on Alpine Linux.

cjwatson committed:
Move PyNaCl to a new conch_nacl extra

cjwatson commented:
@adiroiban Thanks.  I agree that a soft dependency makes sense; for some reason I just hadn't really thought of the approach of adding a new extra.  I hope this is better now.

(I noticed #1605 while looking for possibly-related tickets before filing my own.  This bit has been in my sprawling plan since 2018 ...)

adiroiban commented on src/twisted/conch/ssh/keys.py:
```diff
@@ -26,6 +27,8 @@
     load_pem_private_key,
     load_ssh_public_key,
 )
+from nacl.exceptions import BadSignatureError  # type: ignore[import]
```
I think that for a soft-dependency this needs to be wrapped in a `try/except ImportError` and define `BadSignatureError = object` or something like this

I am no what  is best way to implement it as a soft-dependency.
Because if you have it as a soft-dependency  and pynal is not installed and you have an older OpenSSL  you will not have support for ED25519

cjwatson replied:
Oh, right, good point.  I'll sort this out as soon as I get a moment.

adiroiban replied:
No hurry from my side. I did a quick review and the code looks very well written.
I am looking to do a 20.7.0 release ASAP...but we can also do a 20.8.0 release to include this PR.

glyph committed:
Merge branch 'trunk' into 10208-conch-ed25519-pynacl

cjwatson committed:
Merge branch 'trunk' into 10208-conch-ed25519-pynacl

cjwatson committed:
Pacify mypy

Stitching in nacl.signing using composition rather than inheritance
makes this all much simpler.

cjwatson committed:
Make PyNaCl an optional dependency

This way people with modern OpenSSL and cryptography don't need to worry
about additional build-dependencies for PyNaCl.

cjwatson committed:
Add %(conch_nacl)s to mypy extra

Otherwise mypy can't analyse nacl's types.

cjwatson commented:
@adiroiban Sorry for my long delay on this!  I got badly bogged down in `mypy` issues, until I realized that if I used composition instead of inheritance to stitch in the `nacl.signing`-based implementation then everything got a lot simpler, because that made space for simply inheriting from `ed25519.Ed25519{Public,Private}Key` and then type inference has a much easier time of it.

This is ready for another review now.

cjwatson committed:
Use Ed25519PrivateKey wrapper in ckeygen too

cjwatson committed:
Add a little more test coverage

adiroiban committed:
Merge branch 'trunk' into 10208-conch-ed25519-pynacl

adiroiban reviewed (approved):
Excellent work. Thanks!

I have updated the branch with master to trigger a new round of tests.

If all is green this can be merged. 

You left the commit message in the PR description, so anyone can see this PR green can merge it :)

adiroiban commented:
I have enabled the auto-merge. In this way we should not have to wait for the CI status in order to merge. 

Let me know if you prefer manual merge.
And I think that you can cancel the auto-merge for this PR at any time.

cjwatson committed:
Merge branch 'trunk' into 10208-conch-ed25519-pynacl
